libs/contrib/Text/PrettyPrint/Prettyprinter/Doc.idr

```diff
@@ -0,0 +1,743 @@
+    go (RecordedWithespace lines spaces) (SChar c rest) =
+      do rest' <- go (RecordedWithespace [] 0) rest
+         pure $ commitWhitespace lines spaces (SChar c rest')
+    go (RecordedWithespace lines spaces) (SText l text rest) =
+      let stripped = pack $ dropWhileEnd (== ' ') $ unpack text
+          strippedLength = cast $ length stripped
+          trailingLength = l - strippedLength in
+          if strippedLength == 0
+             then go (RecordedWithespace lines (spaces + l)) rest
+             else do rest' <- go (RecordedWithespace [] trailingLength) rest
+                     pure $ commitWhitespace lines spaces (SText strippedLength stripped rest')
+    go (RecordedWithespace lines spaces) (SLine i rest) = go (RecordedWithespace (i :: lines) 0) rest
+    go (RecordedWithespace lines spaces) (SAnnPush ann rest) =
+      do rest' <- go (AnnotationLevel 1) rest
+         pure $ commitWhitespace lines spaces (SAnnPush ann rest')
+    go (RecordedWithespace lines spaces) (SAnnPop _) = Nothing
+
+public export
+FittingPredicate : Type -> Type
+FittingPredicate ann = Int -> Int -> Maybe Int -> SimpleDocStream ann -> Bool
+
+data LayoutPipeline ann = Nil | Cons Int (Doc ann) (LayoutPipeline ann) | UndoAnn (LayoutPipeline ann)
+
+export
+defaultPageWidth : PageWidth
+defaultPageWidth = AvailablePerLine 80 1
+
+round : Double -> Int
+round x = if x > 0
+             then if x - floor x < 0.5 then cast $ floor x else cast $ ceiling x
+             else if ceiling x - x < 0.5 then cast $ ceiling x else cast $ floor x
+
+||| The remaining width on the current line.
+remainingWidth : Int -> Double -> Int -> Int -> Int
+remainingWidth lineLength ribbonFraction lineIndent currentColumn =
+  let columnsLeftInLine = lineLength - currentColumn
+      ribbonWidth = (max 0 . min lineLength . round) (cast lineLength * ribbonFraction)
+      columnsLeftInRibbon = lineIndent + ribbonWidth - currentColumn in
+      min columnsLeftInLine columnsLeftInRibbon
+
+public export
+record LayoutOptions where
+  constructor MkLayoutOptions
+  layoutPageWidth : PageWidth
+
+export
+defaultLayoutOptions : LayoutOptions
+defaultLayoutOptions = MkLayoutOptions defaultPageWidth
+
+||| The Wadler/Leijen layout algorithm.
+export
+layoutWadlerLeijen : FittingPredicate ann -> PageWidth -> Doc ann -> SimpleDocStream ann
+layoutWadlerLeijen fits pageWidth_ doc = best 0 0 (Cons 0 doc Nil)
+  where
+    initialIndentation : SimpleDocStream ann -> Maybe Int
+    initialIndentation (SLine i _) = Just i
+    initialIndentation (SAnnPush _ s) = initialIndentation s
+    initialIndentation (SAnnPop s) = initialIndentation s
+    initialIndentation _ = Nothing
+
+    selectNicer : Int -> Int -> SimpleDocStream ann -> SimpleDocStream ann -> SimpleDocStream ann
+    selectNicer lineIndent currentColumn x y =
+      if fits lineIndent currentColumn (initialIndentation y) x then x else y
```

Comment:
Here for `selectNicer` to be able to pick `x`, you need to compute the
whole of `y` no matter what. But in practice you only want to know what
the `initialIndentation` is. So making the second argument of `SLine`
lazy may help.

Similarly, for all of the constructors not matched by `initialIndentation`,
you do not actually are what the subdocuments are. So you may want to make
all of these `Lazy` too.

Alternatively, One trick Jean-Philippe Bernardy does in his [A pretty but not
greedy printer](https://dl.acm.org/doi/abs/10.1145/3110250) paper is to be strict in the relevant *sizes* of the document
but lazy in the documents themselves.

All of the usual combinators maintain the invariant that the generated
candidates (it's not greedy so there can be more than one candidate
document at each step) are of the sizes computed strictly. (And, side
note, [it's not too much work](http://agda.github.io/agda-stdlib/Text.Pretty.Core.html) to prove that these invariants are indeed
respected).

I wonder whether you could something like this here too? I.e. have
`Doc` be a pair of some sizes (at least the `initialIndentation` as
we have seen we need it) and some lazy layouts and carefully make
sure that there is no flow of information from the lazy documents
into the sizes so that we may inspect the first without forcing the
second.

Reply:
I've previously tried to add `Lazy` on `Doc` constructors and the `selectNicer`, but the `initialIndentation` suggestion did the trick. Thank you.